builtins: implement session serialization and deserialization

otan opened the pull request:
See individual commits for details.

Resolves #68464

otan committed:
sessiondatapb: move SequenceCache to LocalOnlySessionData

This actually affects the results of my session migration test, so I
think we actually need to include it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68792)
<!-- Reviewable:end -->

stevendanna commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -5862,6 +5862,95 @@ table's zone configuration this will return NULL.`,
+				*evalCtx.SessionData = *sd
+				return tree.MakeDBool(true), nil
+			},
+			Info:       `This function serializes the variables in the current session.`,
```
I think this is Info is for `serialize` not `deserialize`

otan replied:
you sunk my battleship
(good catch)

stevendanna reviewed (approved):
👍 on the testutils changes, seems like a nice helper to have around.

rafiss reviewed (approved):
just some minor nits and questions. feel free to merge when you like!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68792#-:-MhTvtrK3Y52PDmhT5dc:bzpf4u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @stevendanna)

___
*[pkg/sql/session_migration_test.go, line 30 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68792#-MhTuZECAIrWbqeVhp7F:-MhTuZEDBLS70HkHGt9S:b-pug5uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddad043886bbcd2fcf0bd4b8472cd6a302d4e71e/pkg/sql/session_migration_test.go#L30)):*
> ```Go
> )
> 
> func TestSessionMigration(t *testing.T) {
> ```

cool test! maybe add a comment to this function saying what the commands are (reset, exec, dump_vars, compare_vars etc)
___
*[pkg/sql/sem/builtins/builtins.go, line 5906 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68792#-MhTvYr7Ep4JqwekZxIf:-MhTvYr7Ep4JqwekZxIg:b-32pszq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddad043886bbcd2fcf0bd4b8472cd6a302d4e71e/pkg/sql/sem/builtins/builtins.go#L5906)):*
> ```Go
> 					return nil, err
> 				}
> 				return tree.NewDBytes(tree.DBytes(b)), nil
> ```

thoughts on returning this in base64 encoding? maybe it would make it a bit easier to manage on the client side? i don't feel strongly, just something to consider
___
*[pkg/sql/testdata/session_migration/errors, line 30 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68792#-MhTupUnECbIGS-GAx4M:-MhTupUoACqxSTv4spVj:bozs3ti) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddad043886bbcd2fcf0bd4b8472cd6a302d4e71e/pkg/sql/testdata/session_migration/errors#L30)):*
> ```
> pq: crdb_internal.serialize_session(): cannot serialize a session which is inside a transaction
> 
> reset
> ```

could we add another test here showing what happens with prepared statements?

should we just error up front, or migrate the session and ignore them? i think it's worth trying the latter to see if the user experience is good enough

also, even though we aren't doing it right now, curious to hear your thoughts on how we could support prepared statements? maybe we could get to it in stability period.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: implement session serialization and deserialization

Release note (sql change): Implement the
crdb_internal.serialize_session() and
crdb_internal.deserialize_session(bytes) builtins, the former outputs
the session settings in a string that can be deserialized into another
session by the latter.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68792)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @stevendanna)

___
*[pkg/sql/session_migration_test.go, line 30 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68792#-MhTuZECAIrWbqeVhp7F:-MhVaOg16GhSraTvr3ae:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddad043886bbcd2fcf0bd4b8472cd6a302d4e71e/pkg/sql/session_migration_test.go#L30)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

cool test! maybe add a comment to this function saying what the commands are (reset, exec, dump_vars, compare_vars etc)
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 5906 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68792#-MhTvYr7Ep4JqwekZxIf:-MhVaPv49v0Z9aasdW07:b5ofign) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddad043886bbcd2fcf0bd4b8472cd6a302d4e71e/pkg/sql/sem/builtins/builtins.go#L5906)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

thoughts on returning this in base64 encoding? maybe it would make it a bit easier to manage on the client side? i don't feel strongly, just something to consider
</blockquote></details>

i think the `bytes` makes it easy to encode it to whatever, so i'd like to keep it as is.
___
*[pkg/sql/testdata/session_migration/errors, line 30 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68792#-MhTupUnECbIGS-GAx4M:-MhVaXbz8dau7ks227d2:b-m8gxyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddad043886bbcd2fcf0bd4b8472cd6a302d4e71e/pkg/sql/testdata/session_migration/errors#L30)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could we add another test here showing what happens with prepared statements?

should we just error up front, or migrate the session and ignore them? i think it's worth trying the latter to see if the user experience is good enough

also, even though we aren't doing it right now, curious to hear your thoughts on how we could support prepared statements? maybe we could get to it in stability period.
</blockquote></details>

prepared statements are possible but needs a lot of massaging.
prepared portals are definitely a no no.
alternatively, we could re-execute the PREPARE statement on the new connection. not opposed, but not great, hah!
i've blocked this in the second commit.


<!-- Sent from Reviewable.io -->

otan committed:
builtins: prevent serializing prepared statements

Release note: None

otan commented:
bors r=rafiss

rafiss commented:
Thanks for the prepared statement check! does that also block the anonymous prepared statement (one named with `""`)? Perhaps for that we can assume that the client doesn't actually rely on that saved state, so it's safe to migrate the session if that's the only prepared statement.

otan commented:
What is the anonymous prepared statement and how does it get triggered /
work?

On Sat, Aug 21, 2021 at 3:48 PM Rafi Shamim ***@***.***>
wrote:

> Thanks for the prepared statement check! does that also block the
> anonymous prepared statement (one named with "")? Perhaps for that we can
> assume that the client doesn't actually rely on that saved state, so it's
> safe to migrate the session if that's the only prepared statement.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/68792#issuecomment-903063989>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA32FQ6GOPEOFAFUTM6XUMDT5444PANCNFSM5CAUCJ6Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Oliver Tan | Member of Technical Staff
Cockroach Labs

rafiss commented:
In the protocol flow it's common for clients send a Parse/Bind/Execute all at once, without including any name in the Parse command.

But the unnamed prepared statement lasts until the next Parse command. 

> If successfully created, a named prepared-statement object lasts till the end of the current session, unless explicitly destroyed. An unnamed prepared statement lasts only until the next Parse statement specifying the unnamed statement as destination is issued. (Note that a simple Query message also destroys the unnamed statement.) Named prepared statements must be explicitly closed before they can be redefined by another Parse message, but this is not required for the unnamed statement. Named prepared statements can also be created and accessed at the SQL command level, using PREPARE and EXECUTE.

So reading this more closely, it looks like the check you added already covers this.
